### DISCUSSION TITLE
Inspecting a query plan

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#12110] [Usage Questions] [answered]
The Wiki page [Query Plan SQL construct](https://github.com/sqlalchemy/sqlalchemy/wiki/Query-Plan-SQL-construct) doesn’t contain a SQLA v2 snippet, and I was hoping for an update? For now I’ll go with the v1.4 version, which hopefully should just work…

Thank you!

**Comment by CaselIT** [comment]
Hi,

I guess we can update it, but 2.0 would be basically the same as 1.4, but using mapped instead of columns for the orm declaration. Do you want to update?

**Reply by jenstroeger** [reply]
Oh right, I can clone the Wiki repo and submit a PR?

**Comment by zzzeek** [comment]
it works fine as is, there's one warning about inherit cache so just add inherit_cache=False (or True, I went with False) for that

**Reply by zzzeek** [reply, marked as answer]
should be inherit_cache=False.  wiki is updated

**Comment by jenstroeger** [comment]
Thank you @zzzeek and @CaselIT for the quick update!

One question regarding that implementation: I’m seeing an exception
```
  File "/path/to/.venv/lib/python3.10/site-packages/sqlalchemy/sql/base.py", line 278, in _generative
    self = self._generate()
AttributeError: 'explain' object has no attribute '_generate'. Did you mean: '_negate'?
```
from a [`do_orm_execute`](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event handler of ours — would be a check for that attribute be sufficient to exit the event handler early, should I better use an `isinstance()`, or would you need more context on the event handler to give an informed opinion? In this case, it would be ok to just exit the handler, actually, because it wouldn’t be needed for `explain` queries.

**Reply by zzzeek** [reply]
it would be helpful to see the codepath taht calls that `_generate()`, there's lots of kinds of statements that might not have this method

**Reply by jenstroeger** [reply]
The code path contains a bunch of pytest cruft because it’s a failing test, but the relevant traceback is this:
```
  ...

  File "/path/to/.venv/lib/python3.10/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/path/to/tests/test_utilities.py", line 243, in test_explain
    query_plan = dbsession.execute(explain(query)).all()
  File "/path/to/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2362, in execute
    return self._execute_internal(
  File "/path/to/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2207, in _execute_internal
    fn_result: Optional[Result[Any]] = fn(orm_exec_state)
  File "/path/to/src/orm/events.py", line 28, in _add_soft_delete_filtering_criteria
    execute_state.statement = execute_state.statement.options(
  File "<string>", line 2, in options
  File "/path/to/.venv/lib/python3.10/site-packages/sqlalchemy/sql/base.py", line 278, in _generative
    self = self._generate()
AttributeError: 'explain' object has no attribute '_generate'. Did you mean: '_negate'?
```

**Reply by zzzeek** [reply]
OK `.options`, duck-type for that method, not `_generate()`

**Reply by jenstroeger** [reply]
Great, will do:
```python
if not hasattr(execute_state, "options"):
    return
```
Thank you!

**Reply by jenstroeger** [reply]
Actually, other tests fail when I check for `.options` whereas 
```python
if not isinstance(execute_state.statement, explain):
    return
```
as a guard in the event handler passes all tests.

**Comment by jenstroeger** [comment]
As for typing, I used these annotations. Would that be correct?
```python
import sqlalchemy as sa

class explain(sa.sql.expression.Executable, sa.sql.expression.ClauseElement):
    inherit_cache = False

    def __init__(self, stmt: sa.sql.visitors.Visitable, analyze: bool = False) -> None:
        self.statement = stmt
        self.analyze = analyze

    # Lint asks for this method to be implemented, and we don't bother typing it
    # in more detail because it's not being used anyway.
    @property
    def _all_selected_columns(self) -> collections.abc.Sequence[typing.Any]:
        raise NotImplementedError()

@sa.ext.compiler.compiles(explain, "postgresql")
def _pg_explain(element: explain, compiler: sa.sql.compiler.SQLCompiler, **kwargs: typing.ParamSpecKwargs) -> str:
    text = "EXPLAIN "
    if element.analyze:
        text += "ANALYZE "
    text += compiler.process(element.statement, **kwargs)
    return text
```
